compat/c_layer.go

```diff
@@ -1,48 +1,159 @@
 package main
 
-import "fmt"
-import "github.com/google/go-jsonnet"
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
 
-// #cgo CXXFLAGS: -std=c++11 -Wall
-import "C"
+	"github.com/google/go-jsonnet"
 
-//export test
-func test() {
-	fmt.Printf("hello")
+	// #cgo CXXFLAGS: -std=c++11 -Wall
+	// #include "internal.h"
+	"C"
+)
+
+const maxID = 1000
+
+type vm struct {
+	*jsonnet.VM
+	importer *jsonnet.FileImporter
 }
 
-//export run_jsonnet
-func run_jsonnet(code *C.char) {
-	s := C.GoString(code)
-	vm := jsonnet.MakeVM()
-	out, err := vm.EvaluateSnippet("c", s)
-	if err != nil {
-		fmt.Println(err)
+// Because of Go GC, there are restrictions on keeping Go pointers in C.
+// We cannot just pass *jsonnet.VM to C. So instead we use "handle" structs in C
+// which refer to JsonnetVM by a numeric id.
+// The way it is implemented below is simple and has low overhead, but requires us to keep
+// a list of used IDs. This results in a permanent "leak". I don't expect it to ever
+// become a problem.
+var VMs = []*vm{}
```

Comment:
Can you comment that 0 is never a valid id, and the index of VM in the array is always `id - 1`

Reply:
done